File: backend/sonarsentinel/api/filters.py
"""Detection filters, sorting and export scopes (ADR-018 §3–4).

One implementation serves ``GET /surveys/{id}/detections``, ``scope=filtered`` report exports and
the mock server, so the dashboard, the API and downloaded files always select the same detections.
"""

from __future__ import annotations

from collections import Counter
from collections.abc import Callable, Iterable
from dataclasses import dataclass, replace
from typing import Any

from sonarsentinel.errors import ValidationError

CLASSES = ("shipwreck", "pipe", "cylinder", "ghost_net", "debris_other", "unknown_anomaly")
TIERS = ("hazard", "review", "anomaly", "hidden")
REVIEW_STATUSES = ("pending", "confirmed", "rejected", "reclassified")
QUALITY_FLAGS = (
    "DROPOUT",
    "HIGH_MOTION",
    "NEAR_NADIR",
    "SURFACE_RETURN_BAND",
    "TILE_EDGE",
    "GPS_INTERPOLATED",
    "LAYBACK_ESTIMATED",
    "HEADING_FROM_COG",
    "NO_ALTITUDE_BOTTOM_TRACKED",
    "NOT_GEOTAGGED",
)
SORTS = ("confidence", "-confidence", "area", "-area", "ping", "-ping", "class")
SCOPES = ("all", "filtered", "hazards", "confirmed")
DEFAULT_SORT = "-confidence"
MAX_LIMIT = 5000

Detection = dict[str, Any]


@dataclass(frozen=True)
class DetectionQuery:
    """Parsed query parameters; ``None`` means "no restriction"."""

    classes: frozenset[str] | None = None
    min_conf: float | None = None
    max_conf: float | None = None
    tiers: frozenset[str] | None = None
    review_statuses: frozenset[str] | None = None
    flags: frozenset[str] | None = None
    bbox: tuple[float, float, float, float] | None = None
    include_rejected: bool = True
    sort: str = DEFAULT_SORT


def _csv(name: str, value: str | None, allowed: Iterable[str]) -> frozenset[str] | None:
    if value is None:
        return None
    items = frozenset(v.strip() for v in value.split(",") if v.strip())
    if not items:
        return None
    allowed_set = set(allowed)
    unknown = sorted(items - allowed_set)
    if unknown:
        raise ValidationError(
            f"Unknown {name}: {', '.join(unknown)}", field=name, supported=sorted(allowed_set)
        )
    return items


def parse_bbox(value: str | None) -> tuple[float, float, float, float] | None:
    """``minLon,minLat,maxLon,maxLat`` → tuple; 400 for anything else."""
    if value is None or not value.strip():
        return None
    message = "bbox must be minLon,minLat,maxLon,maxLat"
    try:
        parts = [float(v) for v in value.split(",")]
    except ValueError as exc:
        raise ValidationError(message, field="bbox", bbox=value) from exc
    if len(parts) != 4 or parts[0] > parts[2] or parts[1] > parts[3]:
        raise ValidationError(message, field="bbox", bbox=value)
    return parts[0], parts[1], parts[2], parts[3]


def _confidence(name: str, value: float | None) -> float | None:
    if value is None:
        return None
    if not 0.0 <= float(value) <= 100.0:
        raise ValidationError(f"{name} must be between 0 and 100", field=name, value=value)
    return float(value)


def parse_query(
    *,
    cls: str | None = None,
    min_conf: float | None = None,
    max_conf: float | None = None,
    tier: str | None = None,
    review_status: str | None = None,
    flags: str | None = None,
    bbox: str | None = None,
    include_rejected: bool = True,
    sort: str | None = None,
) -> DetectionQuery:
    """Validate the detection query parameters of API spec §2.4 (ADR-018 §3)."""
    low = _confidence("min_conf", min_conf)
    high = _confidence("max_conf", max_conf)
    if low is not None and high is not None and low > high:
        raise ValidationError("min_conf must not exceed max_conf", field="min_conf")
    order = sort or DEFAULT_SORT
    if order not in SORTS:
        raise ValidationError(f"Unknown sort: {order}", field="sort", supported=list(SORTS))
    return DetectionQuery(
        classes=_csv("class", cls, CLASSES),
        min_conf=low,
        max_conf=high,
        tiers=_csv("tier", tier, TIERS),
        review_statuses=_csv("review_status", review_status, REVIEW_STATUSES),
        flags=_csv("flags", flags, QUALITY_FLAGS),
        bbox=parse_bbox(bbox),
        include_rejected=include_rejected,
        sort=order,
    )


def review_status(det: Detection) -> str:
    return str((det.get("review") or {}).get("status", "pending"))


def matches(det: Detection, query: DetectionQuery) -> bool:
    """True when the detection passes every restriction of ``query``."""
    confidence = float(det["confidence"])
    status = review_status(det)
    if query.classes is not None and det["class"] not in query.classes:
        return False
    if query.min_conf is not None and confidence < query.min_conf:
        return False
    if query.max_conf is not None and confidence > query.max_conf:
        return False
    if query.tiers is not None and det["alert_tier"] not in query.tiers:
        return False
    if query.review_statuses is not None and status not in query.review_statuses:
        return False
    if query.flags is not None and not query.flags & set(det.get("quality_flags") or []):
        return False
    if not query.include_rejected and status == "rejected":
        return False
    if query.bbox is not None:
        position = det.get("position") or {}
        lat, lon = position.get("lat"), position.get("lon")
        if lat is None or lon is None:
            return False
        min_lon, min_lat, max_lon, max_lat = query.bbox
        if not (min_lon <= float(lon) <= max_lon and min_lat <= float(lat) <= max_lat):
            return False
    return True


def _sort_key(order: str) -> Callable[[Detection], tuple[Any, ...]]:
    """Sort key with ``detection_id`` as tie-breaker so results are deterministic."""
    field = order.lstrip("-")
    sign = -1.0 if order.startswith("-") else 1.0
    if field == "class":
        return lambda d: (d["class"], d["detection_id"])
    if field == "confidence":
        return lambda d: (sign * float(d["confidence"]), d["detection_id"])
    if field == "area":
        return lambda d: (
            sign * float((d.get("dimensions") or {}).get("area_m2") or 0.0),
            d["detection_id"],
        )
    return lambda d: (
        sign * float((d.get("sonar_ref") or {}).get("ping_start") or 0),
        d["detection_id"],
    )


def filter_detections(items: Iterable[Detection], query: DetectionQuery) -> list[Detection]:
    """Matching detections in their original order."""
    return [d for d in items if matches(d, query)]


def query_detections(items: Iterable[Detection], query: DetectionQuery) -> list[Detection]:
    """Matching detections sorted by ``query.sort``."""
    return sorted(filter_detections(items, query), key=_sort_key(query.sort))


def select_scope(
    items: Iterable[Detection],
    scope: str,
    query: DetectionQuery | None = None,
    *,
    include_rejected: bool = False,
) -> list[Detection]:
    """Detections for a report export (ADR-018 §4), in their original order."""
    if scope not in SCOPES:
        raise ValidationError(f"Unknown scope: {scope}", field="scope", supported=list(SCOPES))
    if scope == "filtered":
        base = replace(query or DetectionQuery(), include_rejected=include_rejected)
        return filter_detections(items, base)
    selected = list(items)
    if scope == "hazards":
        selected = [d for d in selected if d["alert_tier"] == "hazard"]
    elif scope == "confirmed":
        selected = [d for d in selected if review_status(d) in ("confirmed", "reclassified")]
    if not include_rejected:
        selected = [d for d in selected if review_status(d) != "rejected"]
    return selected


def summarize(items: Iterable[Detection]) -> dict[str, Any]:
    """Report ``summary`` block for a list of detections."""
    detections = list(items)
    return {
        "total_detections": len(detections),
        "by_class": dict(sorted(Counter(d["class"] for d in detections).items())),
        "by_tier": dict(sorted(Counter(d["alert_tier"] for d in detections).items())),
    }

File: frontend/src/dashboard/filters.ts
/**
 * Client-side filters for S-02 (ST-093). Semantics match `GET /surveys/{id}/detections`, so the
 * "current filters" report scope returns exactly the rows the map shows:
 * class/tier/review status are membership tests, confidence bounds are inclusive and a detection
 * passes the flag filter when it carries any of the selected flags.
 */
import type { Detection, DetectionQuery, DetectionSort, QualityFlag, ReviewStatus } from '../api/client';
import type { AlertTier, DetectionClass } from '../tokens';

export const ALL_CLASSES: DetectionClass[] = [
  'ghost_net',
  'shipwreck',
  'pipe',
  'cylinder',
  'debris_other',
  'unknown_anomaly',
];
export const ALL_TIERS: AlertTier[] = ['hazard', 'review', 'anomaly', 'hidden'];
export const DEFAULT_TIERS: AlertTier[] = ['hazard', 'review', 'anomaly'];
export const ALL_REVIEW_STATUSES: ReviewStatus[] = ['pending', 'confirmed', 'rejected', 'reclassified'];
export const FILTERABLE_FLAGS: QualityFlag[] = [
  'DROPOUT',
  'HIGH_MOTION',
  'NEAR_NADIR',
  'TILE_EDGE',
  'GPS_INTERPOLATED',
  'LAYBACK_ESTIMATED',
  'SURFACE_RETURN_BAND',
];
const SORTS: DetectionSort[] = ['confidence', '-confidence', 'area', '-area', 'ping', '-ping', 'class'];

export interface Filters {
  classes: DetectionClass[];
  minConf: number;
  maxConf: number;
  tiers: AlertTier[];
  /** Any of these flags; empty = no flag filter. */
  flags: QualityFlag[];
  /** Empty = any review status. */
  reviewStatuses: ReviewStatus[];
  sort: DetectionSort;
}

export const DEFAULT_FILTERS: Filters = {
  classes: ALL_CLASSES,
  minConf: 30,
  maxConf: 100,
  tiers: DEFAULT_TIERS,
  flags: [],
  reviewStatuses: [],
  sort: '-confidence',
};

type NumericKey = 'confidence' | 'area' | 'ping';
const NUMERIC: Record<NumericKey, (d: Detection) => number> = {
  confidence: (d) => d.confidence,
  area: (d) => d.dimensions.area_m2 ?? 0,
  ping: (d) => d.sonar_ref.ping_start ?? 0,
};

export function sortDetections(list: readonly Detection[], sort: DetectionSort): Detection[] {
  const out = [...list];
  if (sort === 'class') {
    return out.sort((a, b) => a.class.localeCompare(b.class) || b.confidence - a.confidence);
  }
  const descending = sort.startsWith('-');
  const key = NUMERIC[(descending ? sort.slice(1) : sort) as NumericKey];
  return out.sort(
    (a, b) => (descending ? key(b) - key(a) : key(a) - key(b)) || a.detection_id.localeCompare(b.detection_id),
  );
}

export function filterDetections(list: readonly Detection[], filters: Filters): Detection[] {
  const classes = new Set<string>(filters.classes);
  const tiers = new Set<string>(filters.tiers);
  const statuses = new Set<string>(filters.reviewStatuses);
  const flags = new Set<string>(filters.flags);
  return list.filter(
    (d) =>
      classes.has(d.class) &&
      d.confidence >= filters.minConf &&
      d.confidence <= filters.maxConf &&
      tiers.has(d.alert_tier) &&
      (statuses.size === 0 || statuses.has(d.review.status)) &&
      (flags.size === 0 || d.quality_flags.some((flag) => flags.has(flag))),
  );
}

export function filterAndSort(list: readonly Detection[], filters: Filters): Detection[] {
  return sortDetections(filterDetections(list, filters), filters.sort);
}

export interface Counts {
  classes: Record<string, number>;
  tiers: Record<string, number>;
}

export function countDetections(list: readonly Detection[]): Counts {
  const counts: Counts = { classes: {}, tiers: {} };
  for (const d of list) {
    counts.classes[d.class] = (counts.classes[d.class] ?? 0) + 1;
    counts.tiers[d.alert_tier] = (counts.tiers[d.alert_tier] ?? 0) + 1;
  }
  return counts;
}

const sameSet = (a: readonly string[], b: readonly string[]) =>
  a.length === b.length && a.every((value) => b.includes(value));

function listParam<T extends string>(params: URLSearchParams, key: string, allowed: readonly T[], fallback: T[]): T[] {
  const raw = params.get(key);
  if (raw === null) return fallback;
  return raw
    .split(',')
    .filter((value): value is T => (allowed as readonly string[]).includes(value));
}

function numberParam(params: URLSearchParams, key: string, fallback: number): number {
  const value = Number(params.get(key));
  return params.has(key) && Number.isFinite(value) ? Math.min(Math.max(value, 0), 100) : fallback;
}

/** Filters from the page URL (shareable view); unknown values are ignored. */
export function filtersFromSearch(params: URLSearchParams): Filters {
  const sort = params.get('sort') as DetectionSort | null;
  return {
    classes: listParam(params, 'class', ALL_CLASSES, DEFAULT_FILTERS.classes),
    minConf: numberParam(params, 'min_conf', DEFAULT_FILTERS.minConf),
    maxConf: numberParam(params, 'max_conf', DEFAULT_FILTERS.maxConf),
    tiers: listParam(params, 'tier', ALL_TIERS, DEFAULT_FILTERS.tiers),
    flags: listParam(params, 'flags', FILTERABLE_FLAGS, []),
    reviewStatuses: listParam(params, 'review_status', ALL_REVIEW_STATUSES, []),
    sort: sort && SORTS.includes(sort) ? sort : DEFAULT_FILTERS.sort,
  };
}

/** Writes non-default filters into a copy of `base`, keeping other parameters such as `job`. */
export function filtersToSearch(filters: Filters, base?: URLSearchParams): URLSearchParams {
  const params = new URLSearchParams(base);
  for (const key of ['class', 'min_conf', 'max_conf', 'tier', 'flags', 'review_status', 'sort']) params.delete(key);
  if (!sameSet(filters.classes, DEFAULT_FILTERS.classes)) params.set('class', filters.classes.join(','));
  if (filters.minConf !== DEFAULT_FILTERS.minConf) params.set('min_conf', String(filters.minConf));
  if (filters.maxConf !== DEFAULT_FILTERS.maxConf) params.set('max_conf', String(filters.maxConf));
  if (!sameSet(filters.tiers, DEFAULT_FILTERS.tiers)) params.set('tier', filters.tiers.join(','));
  if (filters.flags.length) params.set('flags', filters.flags.join(','));
  if (filters.reviewStatuses.length) params.set('review_status', filters.reviewStatuses.join(','));
  if (filters.sort !== DEFAULT_FILTERS.sort) params.set('sort', filters.sort);
  return params;
}

/** The same filters as API query parameters (report scope `filtered`). */
export function toDetectionQuery(filters: Filters): DetectionQuery {
  return {
    class: sameSet(filters.classes, ALL_CLASSES) ? undefined : filters.classes,
    tier: filters.tiers,
    min_conf: filters.minConf,
    max_conf: filters.maxConf,
    review_status: filters.reviewStatuses.length ? filters.reviewStatuses : undefined,
    flags: filters.flags.length ? filters.flags : undefined,
  };
}
